contracts/pair_xyk_sale_tax/src/contract.rs

```diff
@@ -0,0 +1,1426 @@
+use std::convert::TryInto;
+use std::str::FromStr;
+use std::vec;
+
+use cosmwasm_schema::cw_serde;
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    attr, coins, from_binary, to_binary, Addr, BankMsg, Binary, CosmosMsg, Decimal, Decimal256,
+    Deps, DepsMut, Env, Fraction, MessageInfo, QuerierWrapper, Reply, ReplyOn, Response, StdError,
+    StdResult, SubMsg, SubMsgResponse, SubMsgResult, Uint128, Uint256, Uint64, WasmMsg,
+};
+use cw2::set_contract_version;
+use cw20::{Cw20ExecuteMsg, Cw20ReceiveMsg, MinterResponse};
+
+use astroport::asset::{
+    addr_opt_validate, check_swap_parameters, format_lp_token_name, Asset, AssetInfo, CoinsExt,
+    PairInfo, MINIMUM_LIQUIDITY_AMOUNT,
+};
+use astroport::factory::PairType;
+use astroport::generator::Cw20HookMsg as GeneratorHookMsg;
+use astroport::pair::{ConfigResponse, DEFAULT_SLIPPAGE, MAX_ALLOWED_SLIPPAGE};
+use astroport::pair::{
+    CumulativePricesResponse, Cw20HookMsg, ExecuteMsg, InstantiateMsg, PoolResponse, QueryMsg,
+    ReverseSimulationResponse, SimulationResponse, TWAP_PRECISION,
+};
+use astroport::querier::{query_factory_config, query_fee_info, query_supply};
+use astroport::{token::InstantiateMsg as TokenInstantiateMsg, U256};
+use cw_utils::parse_instantiate_response_data;
+
+use crate::error::ContractError;
+use crate::state::{Config, BALANCES, CONFIG};
+use astroport::pair_xyk_sale_tax::{
+    MigrateMsg, SaleTaxConfigUpdates, SaleTaxInitParams, TaxConfigChecked,
+};
+use astroport_pair::state::{Config as XykConfig, CONFIG as XYK_CONFIG};
+
+/// Contract name that is used for migration.
+const CONTRACT_NAME: &str = env!("CARGO_PKG_NAME");
+/// Contract version that is used for migration.
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+/// A `reply` call code ID used for sub-messages.
+const INSTANTIATE_TOKEN_REPLY_ID: u64 = 1;
+
+/// Creates a new contract with the specified parameters in the [`InstantiateMsg`].
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    // Validate asset infos
+    if msg.asset_infos.len() != 2 {
+        return Err(StdError::generic_err("asset_infos must contain exactly two elements").into());
+    }
+    msg.asset_infos[0].check(deps.api)?;
+    msg.asset_infos[1].check(deps.api)?;
+    if msg.asset_infos[0] == msg.asset_infos[1] {
+        return Err(ContractError::DoublingAssets {});
+    }
+
+    let init_params = SaleTaxInitParams::from_binary(msg.init_params.clone())?;
+
+    let config = Config {
+        pair_info: PairInfo {
+            contract_addr: env.contract.address.clone(),
+            liquidity_token: Addr::unchecked(""),
+            asset_infos: msg.asset_infos.clone(),
+            pair_type: PairType::Custom(CONTRACT_NAME.to_string()),
```

Comment:
As Astroport is not going to add this specific pair type in the factory query to get fee config from factory will always fail. I suggest to keep `PairType::Xyk {}` for this pair.


The other path would be to add config for this pair with `is_disabled: true` but this blocks this pair from being incentivized in [Generator](https://github.com/astroport-fi/astroport-core/blob/1bfab32a56aedcf341628b70e457da2687be34a9/contracts/tokenomics/generator/src/contract.rs#L2011-L2013). Don't know why we have this check `pair_config.is_disabled` in generator tho.

Reply:
Ah yeah this was from my original idea to add ut as a custom pair type until speaking with @stefanionescu and he suggested that we should instead migrate the standard xyk pair. So in practice this contract will never be instantiated (unless someone wants to instantiate an instance outside of astroport). I can change it to `PairType::Xyk {}` if you prefer.

Reply:
changed it.